src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -51,78 +50,90 @@ export class BrowserAdapterEventManager {
-        this.deferredEvents = this.deferredEvents.filter(
-            event => this.processEvent(event.eventType, event.eventArgs, true) === false,
-        );
+        const stillDeferred: DeferredEventDetails[] = [];
+        this.deferredEvents.forEach(deferredEvent => {
+            if (this.eventsToApplicationListenersMapping[deferredEvent.eventType]) {
+                try {
+                    this.processEvent(deferredEvent.eventType, deferredEvent.eventArgs);
+                    deferredEvent.deferredResolution.resolveHook(true);
+                } catch (error) {
+                    console.error(error);
+                    deferredEvent.deferredResolution.rejectHook();
+                }
+            } else {
+                stillDeferred.push(deferredEvent);
+            }
+        });
+        this.deferredEvents = stillDeferred;
     }
 
-    private async forwardEventToApplicationListener(
+    private forwardEventToApplicationListener(
         listener: ApplicationListener,
         eventArgs: any[],
     ): Promise<any> {
-        const result = listener(...eventArgs);
+        let result: any;
+        try {
+            result = listener(...eventArgs);
+        } catch (error) {
+            console.error('Error thrown in application listener: ', error);
+            throw error;
+        }
         if (!!result && typeof result.then === 'function') {
             //wrapping application listener promise responses in a 4-minute promise race
             //prevents the service worker going idle before a response is sent
-            await this.promiseFactory.timeout(result, FOUR_MINUTES);
+            return this.promiseFactory.timeout(result, FOUR_MINUTES);
         } else {
             // it is possible that this is the result of a fire and forget listener
             // wrap in 2-minute timeout to ensure it completes during service worker lifetime
-            return this.timeoutFactory.timeoutType === TimeoutType.Alarm
-                ? this.timeoutFactory.createTimeout(
-                      () => Promise.resolve(result),
-                      TWO_MINUTES,
-                      `wrapper-timeout-${Date.now()}`,
-                  )
-                : this.timeoutFactory.createTimeout(() => Promise.resolve(result), TWO_MINUTES);
+            globalThis.setTimeout(() => {
+                Promise.resolve(result);
+            }, TWO_MINUTES);
+            return Promise.resolve(result);
         }
-        return result;
     }
 
-    public processEvent(eventType: string, eventArgs: any[], isDeferred?: boolean): boolean {
-        if (this.eventsToApplicationListenersMapping[eventType]) {
-            this.eventsToApplicationListenersMapping[eventType].forEach(applicationListener => {
-                this.forwardEventToApplicationListener(applicationListener, eventArgs);
-            });
-            return true;
+    private deferEvent(eventDetails: EventDetails): Promise<any> {
```

Comment:
I don't think this has the desired effect in the case where the browser sends us an event we never register a handler for. I think what we'd want in that case would be to time out the promise after 4 minutes, but I think this would just pend the promise indefinitely and get our worker torn down after 5 minutes